flamingo/common.h

```diff
@@ -50,50 +50,303 @@ static inline int find_static_members_in_class(flamingo_t* flamingo, flamingo_sc
+ *
+ * @param env The environment.
+ * @return The parent scope, or NULL if there is no parent scope (stack size < 2).
+ */
 static inline flamingo_scope_t* env_parent_scope(flamingo_env_t* env);
+
+/**
+ * Get the current scope in the environment.
+ *
+ * @param env The environment.
+ * @return The current scope.
+ */
 static inline flamingo_scope_t* env_cur_scope(flamingo_env_t* env);
+
+/**
+ * Attach a scope to the environment's stack.
+ *
+ * This pushes the scope onto the stack.
+ *
+ * @param env The environment.
+ * @param scope The scope to attach.
+ */
 static inline void env_gently_attach_scope(flamingo_env_t* env, flamingo_scope_t* scope);
+
+/**
+ * Detach the current scope from the environment's stack.
+ *
+ * This pops the scope from the stack and returns it.
+ *
+ * @param env The environment.
+ * @return The detached scope.
+ */
 static inline flamingo_scope_t* env_gently_detach_scope(flamingo_env_t* env);
+
+/**
+ * Push a new scope onto the environment's stack.
+ *
+ * This allocates a new scope and attaches it. The new scope inherits the `class_scope` property from the parent.
+ *
+ * @param env The environment.
+ * @return The new scope.
+ */
 static inline flamingo_scope_t* env_push_scope(flamingo_env_t* env);
+
+/**
+ * Pop the current scope from the environment's stack.
+ *
+ * This detaches the scope and decrements its reference count.
+ *
+ * @param env The environment.
+ */
 static inline void env_pop_scope(flamingo_env_t* env);
+
+/**
+ * Find a variable in the environment.
+ *
+ * This searches the scope stack from top to bottom, allowing for shadowing.
+ *
+ * @param env The environment.
+ * @param key The name of the variable to find.
+ * @param key_size The size of the variable name.
+ * @return The variable if found, or NULL if not found.
+ */
 static inline flamingo_var_t* env_find_var(flamingo_env_t* env, char const* key, size_t key_size);
 
 // Scope prototypes.
 
+/**
+ * Allocate a new scope.
+ *
+ * The scope is initialized with a reference count of 1.
+ *
+ * @return The new scope.
+ */
 static inline flamingo_scope_t* scope_alloc(void);
+
+/**
+ * Free a scope.
+ *
+ * This frees the scope and all its variables. It should only be called when the reference count reaches 0.
+ *
+ * @param scope The scope to free.
+ */
 static inline void scope_free(flamingo_scope_t* scope);
+
+/**
+ * Decrement the reference count of a scope.
+ *
+ * If the reference count reaches 0, the scope is freed.
+ *
+ * @param scope The scope to decrement the reference count of. Can be NULL.
+ */
 static inline void scope_decref(flamingo_scope_t* scope);
+
+/**
+ * Add a variable to a scope.
+ *
+ * @param scope The scope to add the variable to.
+ * @param key The name of the variable.
+ * @param key_size The size of the variable name.
+ * @return The new variable.
+ */
 static inline flamingo_var_t* scope_add_var(flamingo_scope_t* scope, char const* key, size_t key_size);
+
+/**
+ * Find a variable in a scope (shallow search).
```

Comment:
```suggestion
 * Find a variable in a scope.
```

nit: There's no equivalent to a "deep" search (that would just be `env_find_var`), so specifying its shallow is perhaps misleading. Maybe in the description you could point to `env_find_var` for "deep" search?